### PR TITLE
fix(sqllab): remove link to sqllab if missing perms

### PR DESCRIPTION
**Pull request by villebro**
### SUMMARY
Currently the "View in SQL Lab" option is available in the Dataset menu in Explore view even when the user doesn't belong to the `sql_lab` role. In addition, SQL Lab is accessible (albeit doesn't work properly) when the necessary permissions are missing.

This PR does the following:
- Removes the "View in SQL Lab" option from the Dataset menu
- Redirects the user to the welcome page when trying to access SQL Lab without necessary permissions.
- Remove "Saved Queries" from Welcome page if user is missing `sql_lab` role

### AFTER
Now when a user that doesn't have the `sql_lab` role will no longer see the "View in SQL Lab" menu item:

image

For admin and users with `sql_lab` role the menu is unchanged:

image

When attempting to access SQL Lab without correct perms via a direct URL, the user gets redirected to the welcome page with an error toast. Also, the welcome page doesn't show the "Saved Queries" section if the user doesn't belong to the `sql_lab` role (we could optionally check the specific perms for saved queries, but maybe we can do that in a follow up if that's really needed):

image

Note that due to the way we're currently caching bootstrap data, toasts don't always appear correctly (they can appear delayed or not get cleared after displaying). I will follow up separately to fix this.

### TESTING INSTRUCTIONS
For testing instructions, please see the following comment in this PR: https://github.com/apache/superset/pull/22566#issuecomment-1370601936

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset/views/core.py [hunk 1]
```diff
@@ -2777,6 +2777,13 @@ def _get_sqllab_tabs(user_id: Optional[int]) -> Dict[str, Any]:
     @expose("/sqllab/", methods=["GET", "POST"])
     def sqllab(self) -> FlaskResponse:
         """SQL Editor"""
+        if not (
+            security_manager.is_admin()
+            or "sql_lab" in (role.name for role in security_manager.get_user_roles())
+        ):
+            flash(__("You do not have access to SQL Lab"), "danger")
+            return redirect("/")
```
Comment (dpgaspar) [inline superset/views/core.py:2785]: This probably makes more sense to fix on security manager, where all roles are created

Comment (villebro) [reply superset/views/core.py:2785]: Good idea, I'll move it over 👍

**Review by kgabryje — APPROVED** [review]
Frontend part looks great to me!

**Comment by villebro** [comment]
For testing I've created the following users on the eph env:

UID: `gamma`
PWD: `gamma`
Test case: directly accessing `/superset/sqllab` redirects to welcome page with an error toast and "View in SQL Lab" is missing in Explore menu:
image

UID: `sqlgamma`
PWD: `sqlgamma`
Test case: SQL Lab works and "View in SQL Lab" is present in Explore menu:
image

File: superset-frontend/src/dashboard/util/permissionUtils.test.ts [hunk 2]
```diff
@@ -22,7 +22,11 @@ import {
 } from 'src/types/bootstrapTypes';
 import { Dashboard } from 'src/types/Dashboard';
 import Owner from 'src/types/Owner';
-import { canUserEditDashboard, isUserAdmin } from './permissionUtils';
+import {
+  canUserAccessSqlLab,
+  canUserEditDashboard,
+  isUserAdmin,
+} from './permissionUtils';
```
Comment (villebro) [inline superset-frontend/src/dashboard/util/permissionUtils.test.ts:29]: TODO: these should probably be moved to a general util outside the dashboard section, but refactoring it seems outside the scope of this PR

File: tests/integration_tests/queries/saved_queries/api_tests.py [hunk 3]
```diff
@@ -157,12 +157,12 @@ def test_get_list_saved_query_gamma(self):
         """
         Saved Query API: Test get list saved query
         """
-        gamma = self.get_user("gamma")
+        user = self.get_user("gamma_sqllab")
```
Comment (villebro) [inline tests/integration_tests/queries/saved_queries/api_tests.py:160]: The `gamma_sqllab` test user has both `Gamma` and `sql_lab` roles.

**Comment by villebro** [comment]
FYI, I've recreated the ephemeral environment with the same test users as previously (same UID/PWD):
- **admin**: the regular admin user
- **gamma**: gamma user with access to Examples database
- **sqlgamma**: gamma user with access to Examples database and sql_lab role

The users that have access to SQL Lab also have a saved query which one should be able to see from the Welcome page

**Review by zhaoyongjie — APPROVED** [review]
image

Test in the ephemeral env and it works as expected. Thanks for lots of refactor codes. LGTM.

File: tests/integration_tests/queries/saved_queries/api_tests.py (hunk 3 above)
Comment (dpgaspar) [reply tests/integration_tests/queries/saved_queries/api_tests.py:160]: did we had to change this user for the tests to pass?

**Review by dpgaspar — APPROVED** [review]

**Review by kgabryje — APPROVED** [review]
Re-tested after latest changes, LGTM!
